nixos/modules/misc/nixpkgs.nix

```
@@ -324,74 +324,152 @@ in
+        constructedByMe =
+          # We set it with default priority and it can not be merged, so if the
+          # pkgs module argument has that priority, it's from us.
+          (lib.modules.mergeAttrDefinitionsWithPrio options._module.args).pkgs.highestPrio
+            == lib.modules.defaultOverridePriority
+          # Although, if nixpkgs.pkgs is set, we did forward it, but we did not construct it.
+            && !opt.pkgs.isDefined;
+      in [
+        (
+          let
+            nixosExpectedSystem =
+              if config.nixpkgs.crossSystem != null
+              then config.nixpkgs.crossSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.crossSystem.config))
+              else config.nixpkgs.localSystem.system or (lib.systems.parse.doubleFromSystem (lib.systems.parse.mkSystemFromString config.nixpkgs.localSystem.config));
+            nixosOption =
+              if config.nixpkgs.crossSystem != null
+              then "nixpkgs.crossSystem"
+              else "nixpkgs.localSystem";
+            pkgsSystem = finalPkgs.stdenv.targetPlatform.system;
+          in {
+            assertion = constructedByMe -> !hasPlatform -> nixosExpectedSystem == pkgsSystem;
+            message = "The NixOS nixpkgs.pkgs option was set to a Nixpkgs invocation that compiles to target system ${pkgsSystem} but NixOS was configured for system ${nixosExpectedSystem} via NixOS option ${nixosOption}. The NixOS system settings must match the Nixpkgs target system.";
+          }
+        )
+        {
+          assertion = constructedByMe -> hasPlatform -> legacyOptionsDefined == [];
+          message = ''
+            Your system configures nixpkgs with the platform parameter${optionalString hasBuildPlatform "s"}:
+            ${hostPlatformLine
+            }${buildPlatformLine
+            }
+            However, it also defines the legacy options:
+            ${concatMapStrings showOptionWithDefLocs legacyOptionsDefined}
+            For a future proof system configuration, we recommend to remove
+            the legacy definitions.
+          '';
+        }
+        {
+          assertion = opt.pkgs.isDefined -> cfg.config == {};
+          message = ''
+            Your system configures nixpkgs with an externally created instance.
+            `nixpkgs.config` options should be passed when creating the instance instead.
+
+            Current value:
+            ${lib.generators.toPretty { multiline = true; } opt.config}
+          '';
+        }
+      ];
+    }
+    (mkIf (cfg.flake.source != null) (mkMerge [
+      (mkIf cfg.flake.setNixPath {
+        environment.etc."nix/inputs/nixpkgs".source = mkDefault cfg.flake.source;
+        nix.nixPath = mkDefault [ "nixpkgs=/etc/nix/inputs/nixpkgs" ];
+      })
+      (mkIf cfg.flake.setFlakeRegistry {
+        nix.registry.nixpkgs.to = mkDefault {
+          type = "path";
+          path = cfg.flake.source;
```

```suggestion
          path = lib.cleanSource cfg.flake.source;
```

Maybe have to do this if the user is using a local mutable input which includes dirs like `.git`

---

I don't believe that this should ever apply: with local paths and flakes you get this:

```
configs/nix » ls -la /nix/store/9qb2r3dqpyn25hyxzm9fys98jpgjrwk0-source
total 43185
dr-xr-xr-x     8 root root       19 Dec 31  1969 .
drwxrwxr-t 26439 root nixbld 109327 Oct  3 01:05 ..
dr-xr-xr-x    13 root root       29 Dec 31  1969 doc
dr-xr-xr-x     4 root root       11 Dec 31  1969 .github
dr-xr-xr-x     6 root root       37 Dec 31  1969 lib
dr-xr-xr-x     3 root root        6 Dec 31  1969 maintainers
dr-xr-xr-x     7 root root       13 Dec 31  1969 nixos
dr-xr-xr-x    19 root root       20 Dec 31  1969 pkgs
-r--r--r--     1 root root    36259 Dec 31  1969 CONTRIBUTING.md
-r--r--r--     1 root root     1097 Dec 31  1969 COPYING
-r--r--r--     1 root root      971 Dec 31  1969 default.nix
-r--r--r--     1 root root     2661 Dec 31  1969 .editorconfig
-r--r--r--     1 root root     2733 Dec 31  1969 flake.nix
-r--r--r--     1 root root      662 Dec 31  1969 .gitattributes
-r--r--r--     1 root root     3133 Dec 31  1969 .git-blame-ignore-revs
-r--r--r--     1 root root      519 Dec 31  1969 .gitignore
-r--r--r--     1 root root      846 Dec 31  1969 .mailmap
-r--r--r--     1 root root     5186 Dec 31  1969 README.md
-r--r--r--     1 root root        6 Dec 31  1969 .version
```

And with non-flakes, you would have to make the path into a string with context, which you would want to fetchGit anyway to make the name correct, among other things.